src/modules/json.wren

```diff
@@ -0,0 +1,245 @@
+    static ABORT_ON_ERROR { 2 }
+
+    static shouldAbort(options) {
+        return (options == JsonOptions.ABORT_ON_ERROR || options == (JsonOptions.ABORT_ON_ERROR | JsonOptions.ESCAPE_SLASHES))
+    }
+}
+
+class JsonError {
+    line { _line }
+    position { _position }
+    message { _message }
+    found { _found }
+
+    construct new(line, pos, message, found) {
+        _line = line
+        _position = pos
+        _message = message
+        _found = found
+    }
+
+    static empty() {
+        return JsonError.new(0, 0, "", false)
+    }
+}
+
+class JsonStream {
+    foreign stream_begin(value)
+    foreign stream_end()
+    foreign next
+    foreign value
+    foreign error_message
+    foreign lineno
+    foreign pos
+    foreign static escapechar(value, options)
+
+    namespace { "com.domeengine.json:" }
+    isNull { namespace + "JSON_NULL" }
+    isString { namespace + "JSON_STRING" }
+    isNumeric { namespace + "JSON_NUMBER" }
+    isBoolTrue { namespace + "JSON_TRUE" }
+    isBoolFalse { namespace + "JSON_FALSE" }
+    isArray { namespace + "JSON_ARRAY" }
+    isArrayEnd { namespace + "JSON_ARRAY_END" }
+    isObject { namespace + "JSON_OBJECT" }
+    isObjectEnd { namespace + "JSON_OBJECT_END" }
+    isDone { namespace + "JSON_DONE" }
+    isError { namespace + "JSON_ERROR" }
+    isInit { namespace + "JSON_INIT" }
+
+    result { _result }
+    error { _error }
+    options { _options }
+    raw { _raw }
+
+    construct new(raw, options) {
+        _result = {}
+        _error = JsonError.empty()
+        _lastEvent = isInit
+        _raw = raw
+        _options = options
+    }
+
+    begin() {
+        stream_begin(_raw)
+        _result = process(next)
+    }
+
+    end() {
+        stream_end()
+    }
+
+    process(event) {
+
+        _lastEvent = event
+
+        if (event == isBoolTrue || event == isBoolFalse) {
+            return (event == isBoolTrue)
+        }
+
+        if (event == isNumeric) {
+            return Num.fromString(this.value)
+        }
+
+        if (event == isString) {
+            return this.value
+        }
+
+        if (event == isNull) {
+            return null
+        }
+
+        if (event == isArray) {
+            var elements = []
+            while (true) {
+                event = next
+                _lastEvent = event
+                if (event == isArrayEnd || event == isDone || event == isError) {
+                    break
+                }
+                elements.add(process(event))
+            }
+            return elements
+        }
+
+        if (event == isObject) {
+            var elements = {}
+            while (true) {
+                event = next
+                _lastEvent = event
+                if (event == isObjectEnd || event == isDone || event == isError) {
+                    break
+                }
+                elements[this.value] = process(next)
+            }
+            return elements
+        }
+
+        if (event == isError) {
```

Comment:
It turns out that this only fires if the event occurs in the top level of an object. If an error is found inside a nested object/list/etc this won't be reached. We should be able to abort and jump out to the error handling case as soon as the error is discovered.

Reply:
In my tests pdjson would try to interpret the json the best it can. So even if the json is invalid it would parse it anyway.

https://github.com/skeeto/pdjson

"streaming" mode. When a JSON "stream" contains multiple JSON objects (optionally separated by JSON whitespace), the default behavior of the parser is to allow the stream to be "reset," and to continue parsing the stream.

Reply:
Example 
```json
{
    "this can make an error":{
        "nested": {
            "error": true // This is invalid json
        }
        // But it will be considered anyway since the streaming support
        "nested2":{
            "error": falsc
        }
    }
}
```

It would be parsed as:

```{this can make an error: {nested: {error: true}}}```

Reply:
Is it possible to support streaming, but also detect errors in the JSON formatting? If not, I think it's better that we disable the streaming option entirely, because silent failure like this isn't good for game development.

The user/developer might make a.mistake in their configuration, and the json module won't detect it at all, leading to avoidable confusion.

Reply:
Let me check the lib for this option. And see what can be done 👍